File: Go/problem030.go
package projecteuler

// #30 Digit fifth powers - Project Euler
// http://projecteuler.net/index.php?section=problems&id=30
import "fmt"

func pow5(x int) int {
	return x * x * x * x * x
}

func P030() {
	ans := 0
	// 354294 = (9^5)*6
	for i := 2; i <= 354294; i++ {
		x := i
		sum_digits := 0
		for x > 0 {
			sum_digits += pow5(x % 10)
			x /= 10
		}
		if i == sum_digits {
			fmt.Println(i)
			ans += i
		}
	}

	fmt.Println(ans)
}

File: Zig/problem030.zig
// zig version 0.11.0-dev
const std = @import("std");
const stdout = std.io.getStdOut().writer();

// #30 Digit fifth powers - Project Euler
// http://projecteuler.net/problem=30
pub fn main() !void {
    var ans: u64 = 0;

    for (2..354294 + 1) |i| {
        const sum: u64 = blk: {
            var s: u64 = 0;
            var x: u64 = i;
            while (x > 0) : (x /= 10) {
                s += pow5(x % 10);
            }
            break :blk s;
        };

        if (i == sum) {
            ans += i;
        }
    }

    try stdout.print("{}\n", .{ans});
}

fn pow5(x: u64) u64 {
    return x * x * x * x * x;
}
